File: src/utils/earTrainingBattleSfx.ts
/**
 * 耳コピバトル魔法演出用 SE（R2 / jazzify-cdn 配信。開発時は Vite `/cdn-proxy` 経由）
 */
type EarTrainingBattleMagicSfxKind =
  | 'fireball'
  | 'snowflake'
  | 'lightning'
  | 'meteor'
  | 'quota';

const CDN_HOST = 'https://jazzify-cdn.com';
const MAGIC_SFX_PATH = '/sfx/ear-training-battle/fire-magic-1.mp3';

const toPlaybackUrl = (absoluteUrl: string): string => {
  if (absoluteUrl.startsWith(CDN_HOST)) {
    return `/cdn-proxy${absoluteUrl.slice(CDN_HOST.length)}`;
  }
  return absoluteUrl;
};

const magicUrl = (): string => toPlaybackUrl(`${CDN_HOST}${MAGIC_SFX_PATH}`);

let sharedAudio: HTMLAudioElement | null = null;

const ensureAudio = (): HTMLAudioElement | null => {
  if (typeof window === 'undefined') {
    return null;
  }
  if (!sharedAudio) {
    const el = new Audio(magicUrl());
    el.preload = 'auto';
    void el.load();
    sharedAudio = el;
  }
  return sharedAudio;
};

/** 魔法エフェクト発火時のみ呼ぶ（詠唱・敵攻撃は除く）。同一音源を種別ごとに共用。 */
export const playEarTrainingBattleMagicSfx = (kind: EarTrainingBattleMagicSfxKind): void => {
  void kind;
  const audio = ensureAudio();
  if (!audio) {
    return;
  }
  try {
    audio.currentTime = 0;
    void audio.play();
  } catch {
    // ブラウザの自動再生制限などは無視
  }
};

File: ios/Jazzify/EarTraining/Audio/EarTrainingBattleSfx.swift
import AVFoundation
import Foundation

/// 魔法エフェクト発火用 SE。バンドル `ear-training-battle/fire-magic-1.mp3` を優先し、無い場合のみ CDN から取得。
enum EarTrainingBattleMagicSfxKind: Sendable {
    case fireball
    case snowflake
    case lightning
    case meteor
    case quotaReached
}

@MainActor
final class EarTrainingBattleSfx {
    static let shared = EarTrainingBattleSfx()

    private static let remoteURL = URL(string: "https://jazzify-cdn.com/sfx/ear-training-battle/fire-magic-1.mp3")!

    private var player: AVAudioPlayer?
    private var cachedLocalURL: URL?
    private var prefetchTask: Task<Void, Never>?

    private init() {}

    /// コードクイズ起動時に呼び、初回再生を早める。
    func prefetch() {
        prefetchTask?.cancel()
        prefetchTask = Task { @MainActor in
            _ = await self.resolveLocalFileURL()
        }
    }

    /// 詠唱・敵攻撃以外の魔法演出で呼ぶ。
    func play(_ kind: EarTrainingBattleMagicSfxKind) {
        _ = kind
        Task { @MainActor in
            guard let url = await self.resolveLocalFileURL() else { return }
            do {
                if self.player == nil {
                    let p = try AVAudioPlayer(contentsOf: url)
                    p.prepareToPlay()
                    self.player = p
                }
                self.player?.currentTime = 0
                self.player?.play()
            } catch {
                // 失敗時は無音のまま
            }
        }
    }

    private func resolveLocalFileURL() async -> URL? {
        if let cachedLocalURL {
            return cachedLocalURL
        }
        if let bundled = Self.bundledMagicURL() {
            cachedLocalURL = bundled
            return bundled
        }
        let local: URL? = await Self.downloadIfNeeded()
        cachedLocalURL = local
        return local
    }

    private nonisolated static func bundledMagicURL() -> URL? {
        Bundle.main.url(forResource: "fire-magic-1", withExtension: "mp3", subdirectory: "ear-training-battle")
    }

    private nonisolated static func downloadIfNeeded() async -> URL? {
        let caches = FileManager.default.urls(for: .cachesDirectory, in: .userDomainMask).first
        guard let root = caches else {
            return nil
        }
        let dir = root.appendingPathComponent("EarTrainingBattleSfx", isDirectory: true)
        let destination = dir.appendingPathComponent("fire-magic-1.mp3", isDirectory: false)
        if FileManager.default.fileExists(atPath: destination.path) {
            return destination
        }
        do {
            try FileManager.default.createDirectory(at: dir, withIntermediateDirectories: true)
            let (data, response) = try await URLSession.shared.data(from: remoteURL)
            guard let http = response as? HTTPURLResponse, (200...299).contains(http.statusCode) else {
                return nil
            }
            try data.write(to: destination, options: .atomic)
            return destination
        } catch {
            return nil
        }
    }
}
